moshi/src/main/java16/com/squareup/moshi/RecordJsonAdapter.java

```diff
@@ -0,0 +1,193 @@
+      return null;
+    }
+
+    var rawType = Types.getRawType(type);
+    if (!rawType.isRecord()) { return null; }
+
+    Map<String, Type> mappedTypeArgs = null;
+    if (type instanceof ParameterizedType parameterizedType) {
+      Type[] typeArgs = parameterizedType.getActualTypeArguments();
+      var typeVars = rawType.getTypeParameters();
+      mappedTypeArgs = new LinkedHashMap<>(typeArgs.length);
+      for (int i = 0; i < typeArgs.length; ++i) {
+        var typeVarName = typeVars[i].getName();
+        var materialized = typeArgs[i];
+        mappedTypeArgs.put(typeVarName, materialized);
+      }
+    }
+    var components = rawType.getRecordComponents();
+    var bindings = new LinkedHashMap<String, ComponentBinding<?>>();
+    var constructorParams = new Class<?>[components.length];
+    for (int i = 0, componentsLength = components.length; i < componentsLength; i++) {
+      RecordComponent component = components[i];
+      constructorParams[i] = component.getType();
+      var name = component.getName();
+      var componentType = component.getGenericType();
+      if (componentType instanceof TypeVariable<?> typeVariable) {
+        var typeVarName = typeVariable.getName();
+        if (mappedTypeArgs == null) {
+          throw new AssertionError("No mapped type arguments found for type '" + typeVarName + "'");
+        }
+        var mappedType = mappedTypeArgs.get(typeVarName);
+        if (mappedType == null) {
+          throw new AssertionError("No materialized type argument found for type '" + typeVarName + "'");
+        }
+        componentType = mappedType;
+      }
+      var jsonName = name;
+      Set<Annotation> qualifiers = null;
+      for (var annotation : component.getDeclaredAnnotations()) {
+        if (annotation instanceof Json jsonAnnotation) {
+          jsonName = jsonAnnotation.name();
+        } else {
+          if (annotation.annotationType().isAnnotationPresent(JsonQualifier.class)) {
+            if (qualifiers == null) {
+              qualifiers = new LinkedHashSet<>();
+            }
+            qualifiers.add(annotation);
+          }
+        }
+      }
+      if (qualifiers == null) {
+        qualifiers = Collections.emptySet();
+      }
+      var adapter = moshi.adapter(componentType, qualifiers);
+      var accessor = component.getAccessor();
+      accessor.setAccessible(true);
+      var componentBinding = new ComponentBinding<>(name, jsonName, adapter, accessor);
+      var replaced = bindings.put(jsonName, componentBinding);
+      if (replaced != null) {
+        throw new IllegalArgumentException(
+            "Conflicting components:\n"
+                + "    "
+                + replaced.name
+                + "\n"
+                + "    "
+                + componentBinding.name);
+      }
+    }
+    Constructor<Object> constructor;
+    try {
+      //noinspection unchecked
+      constructor = (Constructor<Object>) rawType.getDeclaredConstructor(constructorParams);
```

Comment:
Is there no way to get a MethodHandle from the Class once we know it's a record? Can't believe they expect you to use ancient java.lang.reflect reflection to programmatically instantiate these...

Reply:
Honestly never used that API, let me look

Reply:
Good news is we can use MethodHandle for this but bad news is `RecordComponent` itself still returns only `Method` for its accessors :/

Reply:
Constructor bit is in 355422c

I also wrapped the accessor Method with one too, but not sure (lack of familiarity) if this is worthwhile 4700a63. Let me know